File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* res;
        ListNode* p1;
        ListNode* p2;
        res = p1 = p2 = new ListNode(-1, head);

        int c = n+1;

        while(c>0){
            c--;
            p2 = p2->next;
        }
        // cout << p2->val <<' ';
        while(p2){
            p2 = p2->next;
            p1 = p1->next;
        }
        cout << p1->val;
        p1->next = p1->next->next;
        

        
        return res->next;
    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        res = ListNode(0, head)
        curr = head
        last = None
        l = 0
        while True:
            c = k
            begin = curr
            while curr and c > 0:
                curr = curr.next
                c -= 1
            if c != 0:
                break
            else:
                prev = curr
                curr = begin
                for _ in range(k):
                    nxt = curr.next
                    curr.next = prev
                    prev = curr
                    curr = nxt
                if l == 0:
                    res.next = prev
                if last:
                    last.next = prev
                last = begin
            l += 1

        return res.next
            
        